feat: add support for a trusted proxies ingress annotation

lucasrod16 opened the pull request:
Closes #93

This PR adds support for a `caddy.ingress.kubernetes.io/trusted-proxies` ingress annotation

mavimo commented:
@lucasrod16 👋 

Thanks for your contribution!

Can I suggest to expand the test scenario to include IPv6 addresses?

TBH I'm not sure on wrapping stdlib error messages. I know is a quite common pattern but exposing the way we are parsing the IP addresses to the customer I don't think is useful to the end user (see https://github.com/caddyserver/ingress/pull/249/files#diff-483273f30e77c76a1abcb95941ea6ccb7a583fc947d0ad6149fd0043abcc24c6R44 ) WDYT?

lucasrod16 commented:
> @lucasrod16 👋
> 
> Thanks for your contribution!
> 
> Can I suggest to expand the test scenario to include IPv6 addresses?
> 
> TBH I'm not sure on wrapping stdlib error messages. I know is a quite common pattern but exposing the way we are parsing the IP addresses to the customer I don't think is useful to the end user (see https://github.com/caddyserver/ingress/pull/249/files#diff-483273f30e77c76a1abcb95941ea6ccb7a583fc947d0ad6149fd0043abcc24c6R44 ) WDYT?

Hi @mavimo,

Thanks for the feedback!

I've added test cases for IPv6 addresses as suggested. I agree that wrapping stdlib errors might not be helpful for end users. I’ve updated the code to return the stdlib errors directly.

The stdlib errors include function names like `netip.ParsePrefix` and `ParseAddr`. If you prefer, we could provide custom error messages to avoid exposing implementation details. Let me know your preference, and I can make that change. Thanks!

mavimo commented:
@lucasrod16 thanks for adding the tests!

> The stdlib errors include function names like netip.ParsePrefix and ParseAddr. If you prefer, we could provide custom error messages to avoid exposing implementation details. Let me know your preference, and I can make that change.

My suggestion is to use return a custom error message that is not related to the internal implementation details (that may change), but that report the problem we face (fail to parse the IP or ..), WDYT?

lucasrod16 commented:
> My suggestion is to use return a custom error message that is not related to the internal implementation details (that may change), but that report the problem we face (fail to parse the IP or ..), WDYT?

I think that would be a good solution. We would lose some detailed context in some cases, such as `IPv4 field has value >255`, `prefix length out of range`, and `each colon-separated field must have at least one digit`, but is likely not critical to providing a good UX.

lucasrod16 commented:
@mavimo I have updated the code to return a custom error message

mavimo commented:
@lucasrod16 thanks, I think it make more sense from the end user prospective, really appreciated! 🙌

mavimo commented:
@lucasrod16 I'm reviewing the test and I think we may consider to improve a bit how we craf tests for this fnctionality.

Let me try to explain a bit what I mean, at the moment we are testing an internal function that do not guarantee that we are properly managing the functionality from the external user prospective (as example we make a mistake and we delete the function call at:
 * https://github.com/caddyserver/ingress/pull/249/files#diff-f2e0d95332910077fffa583de7f6f1bce44757c6a53469bb84f5d7396ecd26f0R50
the test will not fail but the functionality will not be preserved.

My suggestion is to extend the test (potentially ignoring the "internal implementation") like we do in the internal testing:
 * https://github.com/caddyserver/ingress/blob/edba884bb0d83896643368e05990aee5a9f2adad/internal/caddy/ingress/redirect_test.go#L15

may be a bit more verbose and require some scaffolding but will give us more solid results, WDYT?

PS: I don't want to let you feel like "we never reach to the end" I think your contribution is valuable for all the caddy users and I'll be happy to help if needed 😄

lucasrod16 commented:
@mavimo I think that's a great suggestion. I agree that this feature should be tested more thoroughly. I will work on writing more tests to cover this functionality sometime this week as I have time 🙂

lucasrod16 commented:
@mavimo apologies for the delay. I have updated the tests as suggested in https://github.com/caddyserver/ingress/pull/249#issuecomment-2297116319. Please take a look when you get the chance and let me know what you think.

Thanks!

mavimo commented:
@lucasrod16 hey perfect, thx, I just add some extra test to cover other the missing execution path.. thanks for taking care of that! 🙌

mavimo commented:
@Embraser01 @mholt unless you have something to add on top of my review I'll consider to merge it in the next few days :)

lucasrod16 commented:
@mavimo good catch, I appreciate you adding more test cases. Apologies for the oversight on my part

Embraser01 reviewed (approved):
LGTM, thanks!

mholt commented:
I don't really understand the kubernetes stuff, but quick scan LGTM :+1:

mavimo commented:
Thanks  @lucasrod16 !
